http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPServerChannelInitializer.java

```diff
@@ -75,14 +86,89 @@ public void initChannel(SocketChannel ch) throws Exception {
-            ch.pipeline().addLast("ssl", sslHandler);
+        ChannelPipeline p = ch.pipeline();
+        /*
+         * HTTP2 required ALPN Protocol support. Server is required to have ALPN protocol libraries at class path
+         * If http2 in enabled , Assume required libraries available and HTTP/1 and HTTP2 requests will be handled.
+         */
+        if (listenerConfiguration.isHttp2()) {
+            SslContext sslContext = sslContextMap.get(port);
+            if (sslContextMap.get(port) != null) {
+                /*
+                 * In HTTP2 we have two upgrade handlers for HTTPS and HTTP
+                 * HTTP - Default upgrade handler
+                 * HTTPS - Upgrade during SSL Handshake using Application Layer Protocol negotiation .
+                 */
+                // Configure Upgrade handler for HTTP/2 requests Over TLS
+                configureHttp2TLSPipeline(ch, listenerConfiguration, sslContext);
+            } else {
+                // Configure Upgrade handler for HTTP/2 requests
+                configureHttp2Pipeline(ch, listenerConfiguration);
+            }
+        } else {
+            // Configure Pipeline to handle HTTP/1 requests if HTTP/2 not enabled.
+            if (listenerConfiguration.getSslConfig() != null) {
+                SslHandler sslHandler = new SSLHandlerFactory(listenerConfiguration.getSslConfig()).create();
+                ch.pipeline().addLast("ssl", sslHandler);
+
+            }
+            p.addLast("encoder", new HttpResponseEncoder());
+            configureHTTPPipeline(ch, listenerConfiguration);
         }
+    }
+
+    /**
+     * Configure the pipeline for TLS NPN negotiation to HTTP/2.
+     *
+     * @param ch                    Channel
+     * @param listenerConfiguration Listener Configuration
+     * @param sslContext            Netty http2 ALPN SSL context
+     */
+    private void configureHttp2TLSPipeline(SocketChannel ch, ListenerConfiguration listenerConfiguration, SslContext
+            sslContext) {
+        ChannelPipeline p = ch.pipeline();
+        p.addLast("ssl", sslContext.newHandler(ch.alloc()));
+        p.addLast("http-upgrade", new HTTPProtocolNegotiationHandler(connectionManager, listenerConfiguration));
+
+    }
+
+    /**
+     * Configure the pipeline for a cleartext upgrade from HTTP to HTTP/2.0
+     *
+     * @param ch                    Channel
+     * @param listenerConfiguration Listener Configuration
+     */
+    private void configureHttp2Pipeline(SocketChannel ch, ListenerConfiguration listenerConfiguration) {
+        ChannelPipeline p = ch.pipeline();
+        // Add http2 upgrade decoder and upgrade handler for check http version
+        final HttpServerCodec sourceCodec = new HttpServerCodec();
+        final HttpServerUpgradeHandler.UpgradeCodecFactory upgradeCodecFactory = protocol -> {
+            if (AsciiString.contentEquals(Http2CodecUtil.HTTP_UPGRADE_PROTOCOL_NAME, protocol)) {
```

Comment:
there can be multiple upgrade protocols. So if we return then that logic will break